include/nbl/video/CCUDASharedMemory.h

```diff
@@ -0,0 +1,69 @@
+// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_VIDEO_C_CUDA_SHARED_MEMORY_H_
+#define _NBL_VIDEO_C_CUDA_SHARED_MEMORY_H_
+
+
+#ifdef _NBL_COMPILE_WITH_CUDA_
+
+#include "cuda.h"
+#include "nvrtc.h"
+#if CUDA_VERSION < 9000
+	#error "Need CUDA 9.0 SDK or higher."
+#endif
+
+// useful includes in the future
+//#include "cudaEGL.h"
+//#include "cudaVDPAU.h"
+
+namespace nbl::video
+{
+
+class CCUDASharedMemory : public core::IReferenceCounted
+{
+public:
+    friend class CCUDADevice;
+    CUdeviceptr getDevicePtr() const { return m_params.ptr; }
+
+    struct SCreationParams
+    {
+        size_t            size;
+        uint32_t          alignment;
+        CUmemLocationType location;
+    };
+
+    struct SCachedCreationParams : SCreationParams
+    {
+        size_t granularSize;
+        CUmemGenericAllocationHandle mem;
+        CUdeviceptr ptr;
+        union
+        {
+            void* osHandle;
+            int fd;
+        };
+    };
+
+    const SCreationParams& getCreationParams() const { return m_params; }
+
+    core::smart_refctd_ptr<IGPUBuffer> exportAsBuffer(ILogicalDevice* device, core::bitflag<asset::IBuffer::E_USAGE_FLAGS> usage) const;
+    core::smart_refctd_ptr<IGPUImage>  exportAsImage(ILogicalDevice* device, asset::IImage::SCreationParams&& params) const;
```

Comment:
what guarantee do you have that the images will be the same as a CUDA image?

wouldn't it be better to export an `IGPUImage` from something we know is a CUDA image (and we also can query all its stuff to fill `IGPUImage::SCreationParams`)?

Reply:
ok so my source of confusion was the naming, as the function actually exports a memory allocation and creates an image over it.